File: Tests/BroadcastDiagnosticsTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

@MainActor
final class BroadcastDiagnosticsTests: XCTestCase {
    func testRepeatedScanFactsPreserveFirstTimeAndConnectionHistory() {
        var ring = DiagnosticBreadcrumbRing()
        let first = Date(timeIntervalSince1970: 1)
        ring.append(.connection(.connected), at: first)
        ring.append(.broadcastScan("scan already active (watchdog)"), at: first)
        for index in 2...500 {
            ring.append(.broadcastScan("scan already active (watchdog)"),
                        at: Date(timeIntervalSince1970: Double(index)))
        }
        XCTAssertEqual(ring.entries.count, 2)
        XCTAssertEqual(ring.entries.first?.event, .connection(.connected))
        XCTAssertEqual(ring.entries.last?.date, first)
    }

    func testScanCoalescingDoesNotHideInterveningSignalLoss() {
        var ring = DiagnosticBreadcrumbRing()
        ring.append(.broadcastScan("scan already active (watchdog)"))
        ring.append(.signalFreshness(false))
        ring.append(.broadcastScan("scan already active (watchdog)"))
        XCTAssertEqual(ring.entries.count, 3)
        XCTAssertEqual(ring.entries[1].event, .signalFreshness(false))
    }

    func testStoreExportsActualScanFactsThroughExistingReport() {
        let client = RecordingProgressorClient(kind: .whc06)
        let device = DeviceStore(client: client)
        client.onDiagnostic?(.broadcastScan("scan timed out"))
        XCTAssertEqual(device.diagnosticEntries.last?.event, .broadcastScan("scan timed out"))
        let report = DiagnosticReport.text(from: device.diagnosticEntries)
        XCTAssertTrue(report.contains("Bluetooth scan: scan timed out"))
        XCTAssertFalse(report.contains("Stream start written"))
    }

    func testOnlyBroadcastWatchdogRequestsAreQuietWhileStillReachingClient() {
        for kind in [GaugeKind.whc06, .progressor, .entralpi] {
            let client = RecordingProgressorClient(kind: kind)
            let device = DeviceStore(client: client)
            client.setState(.connected)
            device.startStreaming(cause: .watchdog)
            XCTAssertEqual(client.commands, [.startWeightMeasurement])
            XCTAssertEqual(device.diagnosticEntries.contains {
                $0.event == .streamStartRequested(.watchdog)
            }, !kind.capabilities.isBroadcast)

            device.startStreaming(cause: .manualWake)
            XCTAssertTrue(device.diagnosticEntries.contains {
                $0.event == .streamStartRequested(.manualWake)
            })
        }
    }
}

File: android/app/src/test/kotlin/run/nuri/getagrip/BroadcastDiagnosticsTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.
package run.nuri.getagrip

import kotlin.test.*
import run.nuri.getagrip.ble.ProgressorClientDiagnostic
import run.nuri.getagrip.ble.ProgressorConnectionState
import run.nuri.getagrip.ble.StreamStartCause
import run.nuri.getagrip.engine.GaugeKind
import run.nuri.getagrip.store.DeviceStore
import run.nuri.getagrip.store.DiagnosticBreadcrumb
import run.nuri.getagrip.store.DiagnosticBreadcrumbRing
import run.nuri.getagrip.store.diagnosticTimeline

class BroadcastDiagnosticsTests {
    @Test fun repeatedNoOpsRetainTheOriginalFailureAndElapsedTime() {
        val ring = DiagnosticBreadcrumbRing()
        ring.append(DiagnosticBreadcrumb.BroadcastScan("started (initial)"), at = 1_000.0)
        ring.append(DiagnosticBreadcrumb.BroadcastScan("failed (code 6)"), at = 1_600.0)
        repeat(100) {
            ring.append(DiagnosticBreadcrumb.BroadcastScan("retry pending"), at = 1_601.5 + it)
        }
        assertEquals(3, ring.entries.size)
        assertEquals(1_601.5, ring.entries.last().at)
        val report = ring.entries.diagnosticTimeline()
        assertContains(report, "[+600.0s] Bluetooth scan: failed (code 6)")
        assertContains(report, "[+601.5s] Bluetooth scan: retry pending")
        assertFalse(report.contains("1600"))
    }

    @Test fun broadcastLifecycleFactsReachTheOptionalReportWithoutClaimingAWireWrite() {
        val client = RecordingProgressorClient(GaugeKind.whc06)
        val device = DeviceStore(client = client, scope = inertScope(), clock = FakeClock())
        client.onDiagnostic?.invoke(ProgressorClientDiagnostic.BroadcastScan("failed (code 6)"))
        val text = device.diagnosticEntries.diagnosticTimeline()
        assertContains(text, "Bluetooth scan: failed (code 6)")
        assertFalse(text.contains("written"))
    }

    @Test fun onlyBroadcastWatchdogRequestNoiseIsSuppressed() {
        for (kind in listOf(GaugeKind.whc06, GaugeKind.progressor, GaugeKind.entralpi)) {
            val client = RecordingProgressorClient(kind)
            val device = DeviceStore(client = client, scope = inertScope(), clock = FakeClock())
            client.setState(ProgressorConnectionState.Connected)
            device.startStreaming(StreamStartCause.watchdog)
            device.startStreaming(StreamStartCause.manualWake)
            val requests = device.diagnosticEntries.map { it.event }
                .filterIsInstance<DiagnosticBreadcrumb.StreamStartRequested>()
            assertEquals(if (kind == GaugeKind.whc06) listOf(StreamStartCause.manualWake)
                else listOf(StreamStartCause.watchdog, StreamStartCause.manualWake),
                requests.map { it.cause })
            assertEquals(2, client.commands.size, "Reporting must not change stream requests")
        }
    }

    @Test fun emptyTimelineHasNoMisleadingElapsedTime() {
        assertEquals("", emptyList<run.nuri.getagrip.store.DiagnosticBreadcrumbEntry>().diagnosticTimeline())
    }
}
